lib/cretonne/src/licm.rs

```diff
@@ -15,78 +22,189 @@ pub fn do_licm(func: &mut Function,
+                        insts_rpo.push(inst, &mut inst_pool);
+                    }
+                }
+            }
+        }
+    }
+    // We iterate over all the instructions in reverse postorder. If the instruction is part of a
+    // loop, then we look at its arguments: if they are not defined inside the same loop, then the
+    // instruction is loop-invariant and we hoist it. We don't have to test for child loop because
+    // of the reverse postorder and the immediate hoisting guarantees that later instructions that
+    // use its result value can also be hoisted.
+    for &inst in insts_rpo.as_slice(&inst_pool) {
+        // First which loop is the instruction in
+        let inst_loop = match loop_analysis.base_loop_inst(inst, &func.layout) {
+            None => continue,
+            Some(lp) => lp,
+        };
+        // First we find where is the closest definition of one of the instruction's arguments
+        let closest_def_loop = closest_def_position(func, inst, inst_loop, loop_analysis);
+        // If all the instruction's argument are defined outside the loop then this instruction is
+        // loop-invariant
+        let destination_loop = match closest_def_loop {
+            // If all the arguments are defined outside any loops we move the instruction just
+            // before the outermost loop containing inst_loop.
+            DefPosition::OutsideLoop() => loop_analysis.outermost_loop(inst_loop, None),
+            // In this case we move just before the loop which would be a direct child of lp
+            DefPosition::ParentLoop(lp) => loop_analysis.outermost_loop(inst_loop, Some(lp)),
+            // If an argument is defined inside the current loop, we abort because the instruction
+            // is not loop-invariant.
+            DefPosition::InsideCurrentLoop() => continue,
+        };
+        match has_pre_header(&func.layout,
+                             cfg,
+                             domtree,
+                             loop_analysis.loop_header(destination_loop)) {
+            None => {
+                let loop_header = loop_analysis.loop_header(destination_loop);
+                let (_, jump_inst) =
+                    create_pre_header(loop_header, func, cfg, domtree, loop_analysis);
+                move_inst(&mut func.layout, inst, jump_inst);
+            }
+            // If there is a natural pre-header we insert new instructions just before the
+            // related jumping instruction (which is not necessarily at the end).
+            Some((_, last_jump_inst)) => move_inst(&mut func.layout, inst, last_jump_inst),
+        };
+    }
+}
+
+/// Test whether the given opcode is unsafe to even consider for LICM.
+fn trivially_unsafe_for_licm(opcode: Opcode) -> bool {
+    opcode.is_call() || opcode.is_branch() || opcode.is_terminator() || opcode.is_return() ||
+    opcode.can_trap() || opcode.other_side_effects()
+}
```

Comment:
I neglected to mention this earlier, but in the gvn version, `can_load()` and `can_store()` are handled separately in anticipation of gvn possibly be extended to operate on loads and stores some day. So licm needs to check these somewhere also.

We could probably think about factoring out these concerns into utility functions that could be shared by both, and other things in the future, but I'm ok deferring that for now.